### ISSUE TITLE
Proxies break Promises

### ISSUE THREAD

**Issue by overheadhunter** [alpinejs/alpine#161]
I don't quite know why there are proxies everywhere, but I think it is to register observers to properties that are relevant for bindings.

However, this seems to cause problems when invoking functions on objects that are proxied, which rely on promises:

```
Method Promise.prototype.then called on incompatible receiver [object Object]
    at Proxy.then (<anonymous>)
```

To better illustrate this problem, I've created [this minimal example on jsfiddle](https://jsfiddle.net/anzobhmj/).

So far, the only workaround I know of, is to only call functions on unscoped (i.e. globally scoped) objects such as `window.obj.fnUsingPromises()` instead of `obj.fnUsingPromises()`, which comes at the price of losing all possibilities to use data bindings.

**Comment by Pixelrobin** [comment]
For the record, returning promises from within an object's method works. For me most use cases (api request, etc) wouldn't inline a promise but use a method anyways: https://jsfiddle.net/pgyodjw0/1/

```javascript
class MyObj {
  getBarValue() {
    return new Promise((resolve, reject) => {
      resolve(42);
    });
  }

  bar() {
    this.getBarValue().then(val => {
      console.log("bar", val);
    });
  }
}
```

**Comment by overheadhunter** [comment]
> For the record, returning promises from within an object's method works. For me most use cases (api request, etc) wouldn't inline a promise but use a method anyways:

Interesting, but I don't think it covers all use cases. Sometime it makes sense to run a Promise once and reuse it if it is already resolved. Any idea how to do this without using fields (cause it doesn't seem to work with alpine) or globally-scoped variables (because it is kind of an ugly workaround)?

**Comment by SimoTod** [comment]
@overheadhunter Probably not the answer that you wanted to hear...

It's fixable just adding something like in component.js at line 109
```javascript
                // If property is a Promise, just return it.
                if (target[key] && Promise.resolve(target[key]) === target[key]) return target[key]
```
 but I'm not sure that a promise should be assigned to a variable in the alpine context: It wouldn't be reactive and would give to other user the false impression that it could do something magical.

It seems more sensible to create Promises when you actually use it, either in a different function as suggested by @Pixelrobin or directly in your bar function.

```javascript
  bar() {
    const barValue = new Promise((resolve, reject) => {
      resolve(42);
    });
    barValue.then(val => {
      console.log("bar", val);
    });
  }
```

**Comment by overheadhunter** [comment]
> but I'm not sure that a promise should be assigned to a variable in the alpine context: It wouldn't be reactive and would give to other user the false impression that it could do something magical.

I agree! I don't think promises should be used anywhere near bindings. However, alpine doesn't distinguish whether a property is observed or not. Everything gets proxied (apparently recursively..?).

In my application I have this scenario:

```html
<p x-text="obj.statusText"></p>
<button x-on:click="obj.doSomeLongRunningThing()"></button>
```

Obviously I want the `statusText` property to trigger some listener when it changes. But I do not want `doSomeLongRunningThing()` to be touched in any way that impairs its behaviour.

> It seems more sensible to create Promises when you actually use it, either in a different function as suggested by @Pixelrobin or directly in your bar function.

You do note the difference between your suggestion and my original code, do you? You're creating a _new_ promise every time you invoke `bar()`, while I am reusing the existing one, so it gets evaluated only _once_ per object life cycle. Think of it as a lazy loading pattern.

Example: You do not want to re-initialize a websocket connection on each button click, but you still want to wait for it on the first time.

**Comment by SimoTod** [comment]
I see what you mean but it requires further thinking. The current structure, as you saw, doesn't support it out of the box. The change I proposed in the previous post would fix the issue (https://codepen.io/SimoTod/pen/qBdWxBo), theoretically, but it doesn't feel totally right.

Your assumptions about proxies are pretty much correct. Alpine creates proxies for each object because it's the way it implements the reactivity. Note that a function doesn't return a proxy by itself, the issue is that when you assign a Promise to `this.something`, the Promise is saved into the object scope (Proxied) and, since a Promise is an object, when you read it again because you want to return it, it's proxied.

**Comment by overheadhunter** [comment]
Ok thanks for confirming that I understand this correctly. I'll use a workaround for now until there is some other way.

**Comment by calebporzio** [comment]
Should be fixed with the new reactivity core in v2.1.0
